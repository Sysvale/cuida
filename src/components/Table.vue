<!-- eslint-disable vue/multi-word-component-names -->
<template>
	<table class="table__container">
		<tr class="table__header">
			<th
				v-if="showSelect"
				class="table__select-item"
				:class="resolveHeaderItemClass(0)"
			>
				<cds-checkbox
					id="select-all-rows"
					v-model="selectAll"
					class="table__select-checkbox"
					no-text
					:variant="selectVariant"
					@update:model-value="handleSelectAll"
				/>
			</th>
			<th
				v-for="(field, index) in computedFields"
				:key="index"
				:class="resolveHeaderItemClass(index)"
			>
				<!--
					@slot Slot usado para renderizar itens personalizados para o cabeçalho da tabela. Dados do item referente à coluna podem ser acessados através da propriedade `data`. Os dados do escopo do slot podem ser acessados no formato a seguir: slot-scope={ data }
				-->
				<slot
					name="header-item"
					:data="field"
				>
					{{ field.label }}
				</slot>
			</th>
		</tr>
		<tr
			v-for="(item, itemIndex) in items"
			:key="itemIndex"
			:class="resolveItemClass()"
		>
			<td
				v-if="showSelect"
				class="table__select-item"
				:class="resolveContentItemClass(itemIndex, 0)"
			>
				<cds-checkbox
					:id="`select-row-${itemIndex}`"
					v-model="select[itemIndex]"
					class="table__select-checkbox"
					no-text
					:variant="selectVariant"
					@update:model-value="handleSelectRow"
				/>
			</td>
			<td
				v-for="(field, fieldIndex) in computedFields"
				:key="fieldIndex"
				:class="resolveContentItemClass(itemIndex, fieldIndex)"
			>
				<!--
					@slot Slot usado para renderizar itens personalizados para o conteúdo da tabela. Dados do item referente à linha podem ser acessados através da propriedade `data`, enquanto a key referente à coluna pode ser acessada através da propriedade `field`. Os dados do escopo do slot podem ser acessados no formato a seguir: slot-scope={ `data`, `field`, `rowIndex` e `colIndex` }
				-->
				<slot
					name="table-item"
					:data="item"
					:field="field.key"
					:row-index="itemIndex"
					:col-index="fieldIndex"
				>
					{{ item[field.key] }}
				</slot>
			</td>
		</tr>
	</table>
</template>

<script>
import { startCase, findIndex } from 'lodash';
import CdsCheckbox from './Checkbox.vue';

export default {
	components: {
		CdsCheckbox,
	},

	props: {
		/**
		 * Guarda os itens selecionados da tabela.
		 */
		modelValue: {
			type: Array,
			default: () => ([]),
		},
		/**
		 * Array contendo os itens a serem exibidos na tabela. Os itens devem estar no seguinte formato:
		 *
		 *
		 * ``{ 'exampleKey1': 'exampleValue1', 'exampleKey2': 'exampleValue2' }``
		 *
		 * Com a 'key' sendo a mesma do campo da tabela onde o valor deve ser exibido
		 */
		items: {
			type: Array,
			required: true,
			default: () => ([]),
		},
		/**
		 * Array contendo os campos a serem exibidos no cabeçalho da tabela. Os campos devem estar no seguinte formato:
		 *
		 * ``{ key: 'exampleKey1', label: 'exampleLabel1' }``
		 *
		 * Ou em formato de string simples (``'fieldkey'``). Neste caso, o valor exibido no cabeçalho da tabela será a string convertida para [Start Case](https://en.wikipedia.org/wiki/Letter_case#Stylistic_or_specialised_usage).
		 *
		 * Caso essa propriedade não seja enviada, o cabeçalho da tabela será montado de acordo com as `keys` do primeiro objeto do array enviado através da prop `items`.
		 */
		fields: {
			type: Array,
			default: () => ([]),
		},
		/**
		 * Boolean, informa se o estilo será alterado no hover da linha.
		 */
		hover: {
			type: Boolean,
			default: false,
		},
		/**
		 * Boolean, informa se devem ser exibidas checkboxes para selecionar linhas.
		 */
		showSelect: {
			type: Boolean,
			default: false,
		},
		/**
		 * A variante das checkboxes de seleção. São 10 variantes: 'teal', 'green', 'blue',
		 * 'indigo', 'violet', 'pink', 'red', 'orange', 'amber' e 'dark'.
		 */
		selectVariant: {
			type: String,
			default: 'green',
		},
	},

	data() {
		return {
			selectAll: false,
			select: [],
		};
	},

	computed: {
		computedFields() {
			if (this.fields.length > 0) {
				return this.fields.map(field => {
					if (typeof field === 'object') {
						return field;
					}

					return {
						key: field,
						label: startCase(field),
					};
				});
			}

			return this.items.length > 0
				? Object.keys(this.items[0]).map(key =>{
					return {
						key: key,
						label: startCase(key),
					}
				})
				: [];
		},
	},

	watch: {
		modelValue: {
			handler(newValue, oldValue) {
				if (newValue !== oldValue) {
					if (!newValue.length) {
						this.resetSelect();
						return;
					}

					newValue.forEach((item) => {
						const index = findIndex(this.items, item);
						if (index > -1) {
							this.select[index] = true;
						}
					});
				}
			},
			immediate: true,
		},

		items(newValue, oldValue) {
			if (newValue !== oldValue) {
				this.selectAll = false;
				this.resetSelect();
			}
		},

		select:{
			handler(newValue) {
				const selectedItems = this.items.filter((item, index) => newValue[index]);

				/**
				* Evento que indica que o valor do Select foi alterado
				* @event update:modelValue
				* @type {Event}
				*/
				this.$emit('update:modelValue', selectedItems);
			},
			deep: true,
		},
	},

	methods: {
		resetSelect() {
			for (let index = 0; index < this.items.length; index++) {
				this.select[index] = false;
			}
		},

		handleSelectAll() {
			for (let index = 0; index < this.items.length; index++) {
				this.select[index] = this.selectAll;
			}
		},

		handleSelectRow(value) {
			if (this.selectAll && !value) {
				this.selectAll = false;
			}
		},

		resolveHeaderItemClass(index) {
			if (index !== 0 && index !== (this.fields.length - 1)) {
				return 'table__header-item';
			}

			return index === 0 ? 'table__header-item--first' : 'table__header-item--last';
		},

		resolveContentItemClass(itemIndex, index) {
			if (!(itemIndex === this.items.length - 1)) {
				return 'table__item';
			}

			if (index !== 0 && index !== (this.fields.length - 1)) {
				return 'table__item--common';
			}

			return index === 0 ? 'table__item--first' : 'table__item--last';
		},

		resolveItemClass() {
			return this.hover ? 'table__content--hoverable' : '';
		},
	},
};
</script>

<style lang="scss" scoped>
@import '../assets/sass/tokens.scss';

.table {
	&__container {
		border: 1px solid $n-30;
		border-collapse: separate;
		border-radius: $border-radius-extra-small;
		border-spacing: 0px;
		width: 100%;
	}

	&__select-item {
		width: 50px;
	}

	&__select-checkbox {
		scale: 0.75;
		margin-top: 1px;
	}

	&__header {
		background-color: $n-10;

		&-item {
			@include body-2;
			border-bottom: 1px solid $n-30;
			font-weight: 700;
			padding: spacer(3) spacer(4);
			text-align: inherit;

			&--first {
				border-top-left-radius: $border-radius-extra-small;

				@extend .table__header-item;
			}

			&--last {
				border-top-right-radius: $border-radius-extra-small;

				@extend .table__header-item;
			}
		}
	}

	&__content--hoverable {
		&:hover {
			background-color: rgba($n-10, .7);
		}
	}

	&__item {
		@include body-2;
		border-bottom: 1px solid $n-30;
		font-size: 14px;
		max-width: 400px;
		padding: pa(4);
		vertical-align: top;

		&--first {
			border-bottom: none;
			border-bottom-left-radius: $border-radius-extra-small;

			@extend .table__item;
		}

		&--common {
			border-bottom: none;

			@extend .table__item;
		}

		&--last {
			border-bottom: none;
			border-bottom-right-radius: $border-radius-extra-small;

			@extend .table__item;
		}
	}
}

</style>
