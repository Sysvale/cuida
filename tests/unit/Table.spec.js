import { shallowMount, createLocalVue } from '@vue/test-utils';
import Table from '../../src/components/Table.vue';
import flushPromises from 'flush-promises';

const localVue = createLocalVue();

const fields = ['field1', 'field2', 'field3'];

describe('Table', () => {
	test('component renders correctly', async () => {
		const wrapper = shallowMount(Table, {
			localVue,
			propsData: {
				fields,
				items: [
					{
						field1: 'field 1 content',
						field2: 'field 2 content',
						field3: 'field 3 content',
					},
					{
						field1: 'field 1 content',
						field2: 'field 2 content',
						field3: 'field 3 content',
					},
				],
			},
		});

		await flushPromises();

		expect(wrapper).toMatchSnapshot();
	});

	test('fields are built correctly based on fields prop', async () => {
		const wrapper = shallowMount(Table, {
			localVue,
			propsData: {
				fields: [
					{
						key: 'field1',
						label: 'Field 1',
					},
					'field2',
					'Field3',
				],
				items: [
					{
						field1: 'field 1 content',
						field2: 'field 2 content',
						Field3: 'field 3 content',
					},
					{
						field1: 'field 1 content',
						field2: 'field 2 content',
						Field3: 'field 3 content',
					},
				],
			},
		});

		const fieldsToMatch = [
			{
				key: 'field1',
				label: 'Field 1',
			},
			{
				key: 'field2',
				label: 'Field 2',
			},
			{
				key: 'Field3',
				label: 'Field 3',
			},
		];

		await flushPromises();

		expect(wrapper.vm.computedFields).toMatchObject(fieldsToMatch);
	});

	test('fields are built correctly based on items', async () => {
		const wrapper = shallowMount(Table, {
			localVue,
			propsData: {
				items: [
					{
						field1: 'field 1 content',
						field2: 'field 2 content',
						field3: 'field 3 content',
					},
					{
						field1: 'field 1 content',
						field2: 'field 2 content',
						field3: 'field 3 content',
						field4: 'field 4 content',
					},
				],
			},
		});

		const fieldsToMatch = [
			{
				key: 'field1',
				label: 'Field 1',
			},
			{
				key: 'field2',
				label: 'Field 2',
			},
			{
				key: 'field3',
				label: 'Field 3',
			},
		];

		await flushPromises();

		expect(wrapper.vm.computedFields).toMatchObject(fieldsToMatch);
	});
});
