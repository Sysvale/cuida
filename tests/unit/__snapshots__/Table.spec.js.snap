// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component renders correctly 1`] = `
<table class="table__container">
  <tr class="table__header">
    <th class="table__header-item--first">
      Field 1
    </th>
    <th class="table__header-item">
      Field 2
    </th>
    <th class="table__header-item--last">
      Field 3
    </th>
  </tr>
  <tr>
    <td class="table__item">
      field 1 content
    </td>
    <td class="table__item">
      field 2 content
    </td>
    <td class="table__item">
      field 3 content
    </td>
  </tr>
  <tr>
    <td class="table__item--first">
      field 1 content
    </td>
    <td class="table__item--common">
      field 2 content
    </td>
    <td class="table__item--last">
      field 3 content
    </td>
  </tr>
</table>
`;
